From: jbeulich@suse.com
Subject: netback: fix flipping mode
Patch-mainline: Never, SUSE-Xen specific
References: bsc#996664

A receive SKB's shared info structure gets invalidated when flipping
its data to the guest. Just like clearing frag_list and nr_frags we
also need to clear tx_flags (at the very least the SKBTX_DEV_ZEROCOPY
flag) before handing it to dev_kfree_skb().

--- a/drivers/xen/netback/netback.c
+++ b/drivers/xen/netback/netback.c
@@ -693,8 +693,9 @@ static void net_rx_action(unsigned long
 		   non-linear skbs destined for flipping interfaces) */
 		if (!netif->copying_receiver) {
 			atomic_set(&(skb_shinfo(skb)->dataref), 1);
-			skb_shinfo(skb)->frag_list = NULL;
 			skb_shinfo(skb)->nr_frags = 0;
+			skb_shinfo(skb)->tx_flags = 0;
+			skb_frag_list_init(skb);
 			netbk_free_pages(nr_frags, meta + npo.meta_cons + 1);
 		}
 
